enhancements/security/x509-trust.md

```diff
@@ -0,0 +1,226 @@
+
+Administrators will have per-component configuration locations to adjust the cluster-scoped default.
+As for the cluster-scoped default, they will be able to specify whether the per-component trust bundle is to be used in addition to, or instead of, the cluster-scoped default.
+
+### Non-Goals
+
+This enhancement does not address components with existing X.509 trust-distribution mechanisms.
+Trust for those requests is covered by [the Kubernetes API-server Certificates enhancement](../kube-apiserver/certificates.md) and [the Automated Service CA Rotation enhancement](../automated-service-ca-rotation.md).
+
+This enhancement does not address configuration for non-CA TLS settings like ciphers and allowed TLS versions.
+[The ingress TLS configuration enhancement](../kube-apiserver/tls-config.md) covers those settings for cluster-hosted services.
+Outgoing TLS probably needs similar configuration, but that configuration is decoupled from X.509 trust and should be addressed in a separate enhancement.
+
+## Proposal
+
+### Cluster-scoped Default Trust
+
+There will be a ConfigMap (FIXME: or should this be a new CRD under openshift/api?) named `default-ca-bundle` in the `openshift-config` namespace.
+The ConfigMap will have the following keys:
+
+* `ca-bundle.crt`, containing a PEM-encoded X.509 certificate bundle.
+* `composition`, configuring how the bundle in `ca-bundle.crt` is composed with the component's default trust store.
+    Valid values are:
+    * `union`, in which case components should use the union of the `ca-bundle.crt` and their default trust store.
+        This is convenient for cluster administrators who need to extend the default trust store to support additional CAs, but who wish to delegate default trust store maintenance to the cluster components.
+    * `override`, in which case components should use only `ca-bundle.crt` and ignore their default trust store.
+        This is allows cluster admistrators to take complete control of the trust store for situations where delegating trust maintenance is not permissible.
+        For example, a given component may not need to trust all 100+ CAs in [the default certificate store][mozilla-ca-certificates] if it only uses X.509 trust to verify TLS connections to a few services will stable CAs.
+        Excluding unnecessary CAs from the trust store reduces the risk of being exploited via a compromised CA.
+
+        This behavior is also the default when a component recieves an unrecognized `composition` value.
+
+### Per-component Trust
+
+Each component will define its own well-known ConfigMap(s) in its own namespace, with the same `ca-bundle.crt` and `composition` keys as [the cluster-scoped ConfigMap](#cluster-scoped-default-trust).
+For example, the registry operator uses [`trusted-ca`][registry-configmap-name] in [the `openshift-image-registry` namespace][registry-configmap-namespace].
```

Comment:
> Each component will define its own well-known ConfigMap(s) in its own namespace, with the same `ca-bundle.crt` and `composition` keys as [the cluster-scoped ConfigMap](#cluster-scoped-default-trust).

Operators currently define their own configmap statically. See the [registry operator](https://github.com/openshift/cluster-image-registry-operator/blob/master/manifests/04-ca-trusted.yaml) as an example. If I understand correctly, operators will need to get the value of `composition` from the cluster-scoped configmap and create the operator-scoped configmap programatically. Is my understanding correct?

Reply:
> If I understand correctly, operators will need to get the value of `composition` from the cluster-scoped configmap and create the operator-scoped configmap programatically

Nope, just make a create-only ConfigMap, as described [here][1].  Setting `composition` locally will get clobbered by the network operator using [this logic][2] if you have any of the network-operator labels set, but admins could also unset the network-operator labels and set `composition` manually on the per-component ConfigMap as they see fit.

[1]: https://github.com/openshift/enhancements/commit/d6cd880c480dcc586058747934b9b8b54c2b0678#diff-0a1047c277b83965004f694a11aa059bR101
[2]: https://github.com/openshift/enhancements/commit/d6cd880c480dcc586058747934b9b8b54c2b0678#diff-0a1047c277b83965004f694a11aa059bR147-R149